pkg/sql/catalog/typedesc/type_desc.go

```diff
@@ -984,3 +1012,51 @@ func GetTypeDescriptorClosure(typ *types.T) (map[descpb.ID]struct{}, error) {
 func (desc *Mutable) HasPostDeserializationChanges() bool {
 	return desc.changed
 }
+
+func descBuilderFromTableDesc(descriptor catalog.TableDescriptor) TypeDescriptorBuilder {
+	cols := descriptor.VisibleColumns()
+	typs := make([]*types.T, len(cols))
+	names := make([]string, len(cols))
+	for i, col := range cols {
+		typs[i] = col.GetType()
+		names[i] = col.GetName()
+	}
+	typ := types.MakeLabeledTuple(typs, names)
+	tableID := descriptor.GetID()
+	typeOID := TypeIDToOID(tableID)
+	typ.InternalType.Oid = typeOID
+	typ.TypeMeta = types.UserDefinedTypeMetadata{
+		Name: &types.UserDefinedTypeName{
+			Name: descriptor.GetName(),
+		},
+		Version: uint32(descriptor.GetVersion()),
+	}
+	return NewBuilder(&descpb.TypeDescriptor{
+		Name:             descriptor.GetName(),
+		ID:               descpb.ID(typeOID),
```

Comment:
This feels wrong and a likely source of subtle bugs. For enum and alias types, the id is their descriptor ID.

Reply:
This is deliberate. If we kept the TypeDescriptor's ID the same as the TableDescriptor's, it would overwrite the TableDescriptor in various caches, and vice versa.

Reply:
> This is deliberate. If we kept the TypeDescriptor's ID the same as the TableDescriptor's, it would overwrite the TableDescriptor in various caches, and vice versa.

With this approach, that won't happen until descriptor ID `100000` exists, which is something that can very much happen.

Reply:
I changed this to be the table ID.